InnerEye/ML/lightning_models.py

```diff
@@ -0,0 +1,980 @@
+            segmentation=segmentation,
+            ground_truth=cropped_sample.labels_center_crop,
+            allow_multiple_classes_for_each_pixel=True)[:, 1:]
+        # Number of foreground voxels per class, across all crops
+        foreground_voxels = metrics_util.get_number_of_voxels_per_class(cropped_sample.labels)[:, 1:]
+        # Store Dice and voxel count per sample in the minibatch. We need a custom aggregation logic for Dice
+        # because it can be NaN. Also use custom logging for voxel count because Lightning's batch-size weighted
+        # average has a bug.
+        for i in range(dice_per_crop_and_class.shape[0]):
+            dice = self.train_dice if is_training else self.val_dice
+            dice.update(dice_per_crop_and_class[i, :])
+            voxel_count = self.train_voxels if is_training else self.val_voxels
+            voxel_count.update(foreground_voxels[i, :])
+        # store diagnostics per batch
+        center_indices = cropped_sample.center_indices
+        if isinstance(center_indices, torch.Tensor):
+            center_indices = center_indices.cpu().numpy()
+        if is_training:
+            self.train_diagnostics.append(center_indices)
+        else:
+            self.val_diagnostics.append(center_indices)
+        # if self.train_val_params.in_training_mode:
+        #     # store the sample train patch from this epoch for visualization
+        #     if batch_index == self.example_to_save and self.config.store_dataset_sample:
+        #         _store_dataset_sample(self.config, self.train_val_params.epoch, forward_pass_result,
+        #                               cropped_sample)
+        num_subjects = cropped_sample.image.shape[0]
+        self.log_on_epoch(name=MetricType.SUBJECT_COUNT,
+                          value=num_subjects,
+                          is_training=is_training,
+                          reduce_fx=sum,
+                          sync_dist_op=None)
+
+    def training_or_validation_epoch_end(self, is_training: bool) -> None:
+        """
+        Writes all training or validation metrics that were aggregated over the epoch to the loggers.
+        """
+        dice = list((self.train_dice if is_training else self.val_dice).compute_all())
+        for name, value in dice:
+            self.log(name, value)
+        voxel_count = list((self.train_voxels if is_training else self.val_voxels).compute_all())
+        for name, value in voxel_count:
+            self.log(name, value)
+        super().training_or_validation_epoch_end(is_training=is_training)
+
+
+def get_subject_output_file_per_rank(rank: int) -> str:
+    """
+    Gets the name of a file that will store the per-rank per-subject model outputs.
+    :param rank: The rank of the current model in distributed training.
+    :return: A string like "rank7_metrics.csv"
+    """
+    return f"{SUBJECT_OUTPUT_PER_RANK_PREFIX}{rank}"
+
+
+class ScalarLightning(InnerEyeLightning):
```

Comment:
Would it help the users if we split this file into segmentation_module, classification module, and base module py? Mainly to find the corresponding code more easily.

Reply:
good idea, on it.